### PR TITLE
Restrict metric set operations based on metric types

### PR DESCRIPTION
**Pull request by riemanli**
## TL;DR
This pull request introduces a validation check for frequency histogram metrics to ensure they are only computed on union-only set expressions. It also updates the unit test name of when a reporting set is not accessible to the caller and adds a test case for the new validation check.

## What changed
1. In `MetricsService.kt`, a validation check was added to ensure that frequency histogram metrics are only computed on union-only set expressions. If this condition is not met, an `INVALID_ARGUMENT` error is thrown.
2. The name of the test case `createMetric throws INVALID_ARGUMENT when reporting set is not accessible to caller` was updated to `createMetric throws PERMISSION_DENIED when reporting set is not accessible to caller`.
3. A new test case `createMetric throws INVALID_ARGUMENT when Frequency Histogram metric is computed on non-union-only set expression` was added in `MetricsServiceTest.kt`.
4. Comments were added in `metric.proto` to provide more context about the conditions under which univariate statistics are computed.
 
## Why make this change
This change ensures that frequency histogram metrics are computed correctly and provides more informative error messages to the user. It also improves the clarity of the code by adding explanatory comments.

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)

**Comment by riemanli** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1226** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/1226" target="_blank">Graphite</a> 
    * **PR #1233** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/1233" target="_blank">Graphite</a> 
    * **PR #1184** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/1184" target="_blank">Graphite</a> 
      * **PR #1247** <a href="https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/1247" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/world-federation-of-advertisers/cross-media-measurement/1247?utm_source=stack-comment).

**Review by tristanvuong2021 — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @riemanli)

**Review by tristanvuong2021 — APPROVED** [review]

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: all files reviewed, 1 unresolved discussion (waiting on @riemanli)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfCQ_wZ4Cw64y7x2RjO:-NfCQ_wZ4Cw64y7x2RjP:bxsvo8j) (no related file):*
Why a logged warning (which only Reporting service operators will see)? Shouldn't this be an error in Metric creation, or is there some value in getting a result despite the inaccuracy?

___

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: all files reviewed, 1 unresolved discussion (waiting on @SanjayVas)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfCQ_wZ4Cw64y7x2RjO:-NfCSEXFEhsaqLZUYVu2:bxb3u15) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Why a logged warning (which only Reporting service operators will see)? Shouldn't this be an error in Metric creation, or is there some value in getting a result despite the inaccuracy?
</blockquote></details>

For Frequency Histogram, the result accuracy is not guaranteed if the operation is other than union, so I think it makes sense to raise an error. For Impression Count and Watch Duration, their metric values are correct. It's just that we don't support the variances when operations are not union-only. I can modify the documentation to make this more clear. As for the warning, is there a way to let the user see the stdout warning?

___

**Review by SanjayVas — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: all files reviewed, 1 unresolved discussion (waiting on @riemanli)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfCQ_wZ4Cw64y7x2RjO:-NfCYdEw8PqY7u82lGC_:b-yj26cx) (no related file):*
> For Impression Count and Watch Duration, their metric values are correct. It's just that we don't support the variances when operations are not union-only.

Maybe just document this on the variance output fields then?

> As for the warning, is there a way to let the user see the stdout warning?

An API user can only see the API output, which is either a successful response message or an error status.

___

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @SanjayVas and @tristanvuong2021)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfCQ_wZ4Cw64y7x2RjO:-NfDThoMECwJEHns38mI:bs7wumx) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> For Impression Count and Watch Duration, their metric values are correct. It's just that we don't support the variances when operations are not union-only.

Maybe just document this on the variance output fields then?

> As for the warning, is there a way to let the user see the stdout warning?

An API user can only see the API output, which is either a successful response message or an error status.
</blockquote></details>

Done. 

___

**Review by SanjayVas — APPROVED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: all files reviewed, 2 unresolved discussions (waiting on @riemanli)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/MetricsService.kt` line 1400 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfDd_K88maVWeTWKRXI:-NfDd_K88maVWeTWKRXJ:b29fpf4) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fb319f3ab2bae9a2f6f5b935fea221e08db4366/src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/MetricsService.kt#L1400)):*
> ```Kotlin
>   companion object {
>     private val RESOURCE_ID_REGEX = Regex("^[a-z]([a-z0-9-]{0,61}[a-z0-9])?$")
>     private val logger: Logger = Logger.getLogger(this::class.java.name)
> ```

nit: unused
___
*[`src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto` line 204 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfDe8n_5Ia77nHMg0Ea:-NfDe8n_5Ia77nHMg0Eb:b5rnxwc) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fb319f3ab2bae9a2f6f5b935fea221e08db4366/src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto#L204)):*
> ```proto
>     // Univariate statistics of the impression value above.
>     //
>     // If any CMMS impression measurements used in the metric don't have noise
> ```

nit: can probably simplify this by saying "Only set when all of the set operations for the result are unions and all source`Measurement`s contain sufficient information to calculate univariate statistics."

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/MetricsService.kt` line 1400 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfDd_K88maVWeTWKRXI:-NfDilHPCu9nDCsYLmFE:b8ykr3y) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fb319f3ab2bae9a2f6f5b935fea221e08db4366/src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/MetricsService.kt#L1400)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

nit: unused
</blockquote></details>

Thanks for pointing it out. Removed.
___
*[`src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto` line 204 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfDe8n_5Ia77nHMg0Ea:-NfDjD_u378K8kqQESPJ:b-22mamj) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fb319f3ab2bae9a2f6f5b935fea221e08db4366/src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto#L204)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

nit: can probably simplify this by saying "Only set when all of the set operations for the result are unions and all source`Measurement`s contain sufficient information to calculate univariate statistics."
</blockquote></details>

Thanks for the suggestion! Fixed.

**Review by SanjayVas — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: all files reviewed, 1 unresolved discussion (waiting on @Marco-Premier and @riemanli)

___
*[`src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto` line 204 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfDe8n_5Ia77nHMg0Ea:-NfDoT9S0lC0kB9IdZy6:b-wajnp5) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fb319f3ab2bae9a2f6f5b935fea221e08db4366/src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto#L204)):*
<details><summary><i>Previously, riemanli (Rieman) wrote…</i></summary><blockquote>

Thanks for the suggestion! Fixed.
</blockquote></details>

Can use similar phrasing (without the restriction on set operations) for the stats fields in other result types.

**Review by riemanli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: 2 of 3 files reviewed, all discussions resolved (waiting on @Marco-Premier and @SanjayVas)

___
*[`src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto` line 204 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247#-NfDe8n_5Ia77nHMg0Ea:-NfEUnt5-GhQNU0OXivJ:b-896fja) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fb319f3ab2bae9a2f6f5b935fea221e08db4366/src/main/proto/wfa/measurement/reporting/v2alpha/metric.proto#L204)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Can use similar phrasing (without the restriction on set operations) for the stats fields in other result types.
</blockquote></details>

Done!

**Review by Marco-Premier — APPROVED** [review]
> Reviewed 1 of 3 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/1247)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @SanjayVas)
